File: src/events.rs
use poise::serenity_prelude as serenity;

use crate::{
    commands::download_sayas_files,
    config::{access_allowed, Data},
    util::{attach_name, cap_file_body, strip_sgr},
    vm::linked_user,
    webhook::{handle_delete, is_own_message, is_posted_message, post_message},
    Error,
};

fn is_boo_message(s: &str) -> bool {
    s.to_lowercase()
        .split(|c: char| !c.is_alphanumeric())
        .any(|w| w == "boo")
}

const TAUNT: &str = "purged your message haha";

fn blocked_reply_action(
    ref_author: Option<u64>,
    ref_id: Option<serenity::MessageId>,
    ref_content: &str,
    me: u64,
) -> (bool, bool) {
    let mine = matches!(ref_author, Some(a) if a == me)
        || ref_id.map(|id| is_posted_message(&id)).unwrap_or(false);
    match mine {
        true => (true, ref_content != TAUNT),
        false => (false, false),
    }
}
fn artixy_text(s: &str) -> Option<String> {
    let text = s.trim_end().strip_suffix(".ar")?.trim();
    if text.is_empty() {
        return None;
    }
    Some(text.to_string())
}

pub(crate) async fn event_handler(
    ctx: &serenity::Context,
    event: &serenity::FullEvent,
    _framework: poise::FrameworkContext<'_, Data, Error>,
    data: &Data,
) -> Result<(), Error> {
    if let serenity::FullEvent::MessageDelete {
        deleted_message_id,
        ..
    } = event
    {
        let war = data.settings.read().await.war_mode;
        handle_delete(&ctx.http, *deleted_message_id, war).await;
        return Ok(());
    }
    if let serenity::FullEvent::TypingStart { event } = event {
        let name = event
            .member
            .as_ref()
            .and_then(|m| m.nick.clone().or_else(|| m.user.global_name.clone()).or_else(|| Some(m.user.name.clone())))
            .or_else(|| {
                ctx.cache
                    .user(event.user_id)
                    .map(|u| u.global_name.clone().unwrap_or_else(|| u.name.clone()))
            })
            .unwrap_or_else(|| event.user_id.get().to_string());
        let channel_name = ctx
            .cache
            .channel(event.channel_id)
            .map(|c| c.name.clone())
            .unwrap_or_default();
        let _ = crate::feed::log_typing(&name, event.channel_id.get(), &channel_name).await;
        return Ok(());
    }
    let serenity::FullEvent::Message { new_message } = event else {
        return Ok(());
    };
    let channel_name = ctx
        .cache
        .channel(new_message.channel_id)
        .map(|c| c.name.clone())
        .unwrap_or_default();
    let _ = crate::feed::log_message(
        &new_message.author.name,
        new_message.author.bot,
        new_message.channel_id.get(),
        &channel_name,
        &new_message.content,
    )
    .await;
    let id = new_message.author.id.get();
    let (owner, blocked, war) = {
        let a = data.allowed.read().await;
        let s = data.settings.read().await;
        (
            id == a.owner || a.admins.contains(&id),
            a.blocked.contains(&id),
            s.war_mode,
        )
    };
    if blocked && war {
        let me = match ctx.http.get_current_user().await {
            Ok(u) => u.id.get(),
            Err(_) => return Ok(()),
        };
        let (delete, taunt) = match &new_message.referenced_message {
            Some(r) => blocked_reply_action(Some(r.author.id.get()), Some(r.id), &r.content, me),
            None => match &new_message.message_reference {
                Some(r) => match r.message_id {
                    Some(mid) => match new_message.channel_id.message(&ctx.http, mid).await {
                        Ok(orig) => blocked_reply_action(
                            Some(orig.author.id.get()),
                            Some(orig.id),
                            &orig.content,
                            me,
                        ),
                        Err(_) => (false, false),
                    },
                    None => (false, false),
                },
                None => (false, false),
            },
        };
        if delete {
            match new_message.delete(&ctx.http).await {
                Ok(_) => {
                    if taunt {
                        let _ = post_message(&ctx.http, new_message.channel_id, TAUNT.into(), Vec::new()).await;
                    }
                }
                Err(e) => {
                    eprintln!(
                        "guard: failed to delete blocked reply {} in {}: {}",
                        new_message.id.get(),
                        new_message.channel_id.get(),
                        e
                    );
                }
            }
        }
        return Ok(());
    }
    let me_id: u64 = ctx.cache.current_user().id.get();
    if is_own_message(new_message.author.id, new_message.id, serenity::UserId::new(me_id)) {
        return Ok(());
    }
    let mentioned = new_message.mentions.iter().any(|u| u.id.get() == me_id)
        || new_message.content.contains(&format!("<@{me_id}>"))
        || new_message.content.contains(&format!("<@!{me_id}>"))
        || crate::ai::mentions_name(&new_message.content);
    if mentioned {
        let prompt0 = crate::ai::strip_name(&crate::ai::strip_mention(&new_message.content, me_id));
        let is_command = prompt0.starts_with('/') || prompt0.starts_with(';');
        if !is_command {
            let blocked = {
                let a = data.allowed.read().await;
                a.blocked.contains(&id)
            };
            if blocked {
                return Ok(());
            }
            let (ai_on, ai_model, ai_host, ai_key) = {
                let s = data.settings.read().await;
                (
                    s.ai_enabled,
                    s.ai_model.clone(),
                    crate::ai::resolve_host(&s.ollama_host),
                    crate::ai::resolve_ollama_key(&s.ollama_api_key),
                )
            };
            if !ai_on {
                let _ = new_message
                    .reply(&ctx.http, "AI is off — the owner runs `/ai true model:<name>` to enable me.")
                    .await;
                return Ok(());
            }
            let mut prompt = prompt0.clone();
            let display = new_message
                .member
                .as_ref()
                .and_then(|m| m.nick.clone())
                .or_else(|| new_message.author.global_name.clone())
                .unwrap_or_else(|| new_message.author.name.clone());
            let speaker = if display == new_message.author.name {
                display
            } else {
                format!("{display} (@{})", new_message.author.name)
            };
            if prompt.trim().is_empty() {
                if let Some(r) = new_message.referenced_message.as_ref() {
                    let qdisplay = r
                        .author
                        .global_name
                        .clone()
                        .unwrap_or_else(|| r.author.name.clone());
                    let qtext = r.content.trim().to_string();
                    if !qtext.is_empty() {
                        prompt = format!(
                            "(quoting {}): {qtext}",
                            crate::ai::speaker_tag(&qdisplay)
                        );
                    }
                }
            }
            if prompt.trim().is_empty() {
                let _ = new_message
                    .reply(&ctx.http, format!("Ping me with a question — `@artixy <question>` or `artixy <question>` (model `{ai_model}`)."))
                    .await;
                return Ok(());
            }
            if prompt.chars().count() > 4000 {
                prompt = prompt.chars().take(4000).collect();
            }
            let _ = new_message.channel_id.broadcast_typing(&ctx.http).await;
            match crate::ai::ollama_chat(&ai_host, &ai_model, &ai_key, new_message.channel_id.get(), &speaker, &prompt).await {
                Ok(text) => {
                    let chunks = crate::ai::chunk_reply(&text);
                    let mut first = true;
                    for c in chunks {
                        if first {
                            let _ = new_message.reply(&ctx.http, &c).await;
                            first = false;
                        } else if post_message(&ctx.http, new_message.channel_id, c, Vec::new()).await.is_none() {
                            break;
                        }
                    }
                }
                Err(e) => {
                    eprintln!("ai chat failed (model {ai_model} on {ai_host}): {e}");
                    let text = if crate::ai::is_api_full_err(&e.to_string()) {
                        crate::ai::api_full_message()
                    } else {
                        crate::ai::glitch_text(&ai_host, &ai_model).await
                    };
                    let _ = new_message.reply(&ctx.http, &text).await;
                }
            }
            return Ok(());
        }
    }
    if new_message.attachments.is_empty() && !new_message.content.trim().is_empty() {
        let refd_id = new_message
            .referenced_message
            .as_ref()
            .map(|r| r.id)
            .or(new_message.message_reference.as_ref().and_then(|r| r.message_id));
        if let Some(target_id) = refd_id {
            let live_hit = {
                let m = data.live.lock().await;
                m.iter()
                    .find(|(_, e)| e.msg_id == target_id)
                    .map(|(_, e)| (e.author_id, e.in_f.clone(), e.channel))
            };
            if let Some((owner_id, fifo_opt, live_channel)) = live_hit {
                if owner_id.get() != id {
                    let _ = new_message.delete(&ctx.http).await;
                    let dm_text = "That live session belongs to someone else — typing into it is blocked.";
                    if let Ok(dm) = new_message.author.create_dm_channel(&ctx.http).await {
                        let _ = dm
                            .send_message(
                                &ctx.http,
                                serenity::CreateMessage::new().content(dm_text),
                            )
                            .await;
                    }
                    return Ok(());
                }
                let Some(fifo) = fifo_opt else {
                    let _ = new_message.delete(&ctx.http).await;
                    if let Ok(dm) = new_message.author.create_dm_channel(&ctx.http).await {
                        let _ = dm
                            .send_message(
                                &ctx.http,
                                serenity::CreateMessage::new()
                                    .content("That live session isn't interactive (fifo missing)."),
                            )
                            .await;
                    }
                    return Ok(());
                };
                let authed = {
                    let a = data.allowed.read().await;
                    access_allowed(a.owner, &a.users, &a.blocked, id)
                };
                if !authed {
                    return Ok(());
                }
                let trimmed = new_message.content.trim_end();
                if trimmed.is_empty() {
                    return Ok(());
                }
                let payload = crate::live::expand_typed_input(trimmed);
                let runas = linked_user(data, id).await;
                let vm = data.vm.read().await.clone();
                let ok =
                    crate::live::forward_terminal_input(&vm, &fifo, runas.as_deref(), &payload).await;
                let _ = new_message.delete(&ctx.http).await;
                if !ok {
                    if let Ok(dm) = new_message.author.create_dm_channel(&ctx.http).await {
                        let _ = dm
                            .send_message(
                                &ctx.http,
                                serenity::CreateMessage::new()
                                    .content("Couldn't type into that session (it just ended)."),
                            )
                            .await;
                    }
                }
                let _ = live_channel;
                return Ok(());
            }
        }
    }
    if !owner {
        if war && is_boo_message(&new_message.content) {
            let _ = new_message.reply(&ctx.http, "boo on you! :3").await;
        }
        return Ok(());
    }
    if data.settings.read().await.sayas_enabled {
        let content = new_message.content.trim().to_string();
        let has_files = !new_message.attachments.is_empty();
        let plain_text = !content.is_empty()
            && !content.starts_with('/')
            && !content.starts_with(';')
            && !content.ends_with(".ar");
        if plain_text || (has_files && content.is_empty()) {
            let (mut files, _) = download_sayas_files(&new_message.attachments).await;
            let mut body = content.clone();
            if body.chars().count() > 2000 {
                files.insert(
                    0,
                    (
                        attach_name(&body),
                        cap_file_body(&strip_sgr(&body)).into_bytes(),
                    ),
                );
                body = String::new();
            }
            if body.is_empty() && files.is_empty() {
                return Ok(());
            }
            let _ = new_message.delete(&ctx.http).await;
            crate::ai::record_artixy(new_message.channel_id.get(), &body);
            if body.chars().count() <= 2000 && !body.is_empty() && files.is_empty() {
                match &new_message.referenced_message {
                    Some(target) => {
                        let _ = target.reply(&ctx.http, &body).await;
                    }
                    None => {
                        let _ = post_message(&ctx.http, new_message.channel_id, body, Vec::new()).await;
                    }
                }
            } else {
                match &new_message.referenced_message {
                    Some(target) => {
                        let mut builder = serenity::CreateMessage::new()
                            .reference_message((new_message.channel_id, target.id));
                        if !body.is_empty() {
                            builder = builder.content(&body);
                        }
                        for (name, bytes) in files {
                            builder = builder.add_file(serenity::CreateAttachment::bytes(bytes, name));
                        }
                        let _ = new_message
                            .channel_id
                            .send_message(&ctx.http, builder)
                            .await;
                    }
                    None => {
                        let _ = post_message(&ctx.http, new_message.channel_id, body, files).await;
                    }
                }
            }
            return Ok(());
        }
    }
    let Some(text) = artixy_text(&new_message.content) else {
        return Ok(());
    };
    let mut deleted = false;
    for _ in 0..3 {
        match new_message.delete(&ctx.http).await {
            Ok(_) => {
                deleted = true;
                break;
            }
            Err(e) => {
                eprintln!(
                    "artixy-say: delete attempt failed for {} in {}: {}",
                    new_message.id.get(),
                    new_message.channel_id.get(),
                    e
                );
                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
            }
        }
    }
    if !deleted {
        eprintln!(
            "artixy-say: giving up deleting {} (needs Manage Messages there)",
            new_message.id.get()
        );
    }
    let (mut ar_files, _) = download_sayas_files(&new_message.attachments).await;
    let mut ar_body = text.clone();
    if ar_body.chars().count() > 2000 {        ar_files.insert(
            0,
            (
                attach_name(&ar_body),
                cap_file_body(&strip_sgr(&ar_body)).into_bytes(),
            ),
        );
        ar_body = String::new();
    }
    crate::ai::record_artixy(new_message.channel_id.get(), &ar_body);
    if ar_body.chars().count() <= 2000 && !ar_body.is_empty() && ar_files.is_empty() {
        match &new_message.referenced_message {
            Some(target) => {
                let _ = target.reply(&ctx.http, &ar_body).await;
            }
            None => {
                let _ = post_message(&ctx.http, new_message.channel_id, ar_body, Vec::new()).await;
            }
        }
    } else if ar_body.is_empty() && ar_files.is_empty() {
    } else {
        match &new_message.referenced_message {
            Some(target) => {
                let mut builder = serenity::CreateMessage::new()
                    .reference_message((new_message.channel_id, target.id));
                if !ar_body.is_empty() {
                    builder = builder.content(&ar_body);
                }
                for (name, bytes) in ar_files {
                    builder = builder.add_file(serenity::CreateAttachment::bytes(bytes, name));
                }
                let _ = new_message
                    .channel_id
                    .send_message(&ctx.http, builder)
                    .await;
            }
            None => {
                let _ = post_message(&ctx.http, new_message.channel_id, ar_body, ar_files).await;
            }
        }
    }
    Ok(())
}

File: src/events.c
/* Ambient message pipeline (ported 1:1 from events.rs). */
#include "events.h"
#include "ai.h"
#include "bot.h"
#include "commands.h"
#include "discord.h"
#include "live.h"
#include "util.h"
#include "vm.h"

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static const char *taunt = "purged your message haha";

bool events_is_boo(const char *s) {
    if (!s)
        return false;
    const char *p = s;
    while (*p) {
        while (*p && !(( *p >= 'a' && *p <= 'z') ||
                       (*p >= 'A' && *p <= 'Z') ||
                       (*p >= '0' && *p <= '9')))
            p++;
        if (!*p)
            break;
        const char *w = p;
        while (*p && ((*p >= 'a' && *p <= 'z') ||
                      (*p >= 'A' && *p <= 'Z') || (*p >= '0' && *p <= '9')))
            p++;
        if ((size_t)(p - w) == 3 &&
            (w[0] == 'b' || w[0] == 'B') && (w[1] == 'o' || w[1] == 'O') &&
            (w[2] == 'o' || w[2] == 'O'))
            return true;
    }
    return false;
}

char *events_artixy_text(const char *content) {
    if (!content)
        return NULL;
    /* trim trailing whitespace */
    size_t n = strlen(content);
    while (n > 0 && (content[n - 1] == ' ' || content[n - 1] == '\t' ||
                     content[n - 1] == '\n' || content[n - 1] == '\r'))
        n--;
    /* NOTE: strncmp, not strcmp — trailing whitespace beyond n is untouched */
    if (n < 3 || strncmp(content + n - 3, ".ar", 3) != 0)
        return NULL;
    /* trim the suffix + surrounding space */
    size_t m = n - 3;
    while (m > 0 && (content[m - 1] == ' ' || content[m - 1] == '\t'))
        m--;
    if (!m)
        return NULL;
    char *out = xmalloc(m + 1);
    if (!out)
        return NULL;
    memcpy(out, content, m);
    out[m] = '\0';
    return out;
}

/* Post body+files as artix, honouring a reply target (0 = none). */
static void post_as_artix(discord_client_t *dc, uint64_t channel_id,
                          uint64_t reply_to, const char *body,
                          disc_file_t *files, size_t n_files) {
    const char *text = body ? body : "";
    if (reply_to)
        discord_send_reply(dc, channel_id, reply_to, text, files, n_files,
                           NULL);
    else
        discord_send_message(dc, channel_id, text, files, n_files, NULL);
}

void events_handle_message(discord_client_t *c, bot_state_t *st,
                           const disc_message_t *m) {
    if (m->from_webhook)
        return;

    uint64_t me = discord_bot_id(c);
    bool blocked = bot_is_blocked(st, m->author_id);
    bool owner_admin = bot_is_owner(st, m->author_id) ||
                       bot_is_elevated(st, m->author_id);
    bool war = false, sayas_auto = false;
    {
        pthread_rwlock_rdlock(&st->mu);
        war = st->war_mode;
        sayas_auto = st->sayas_enabled;
        pthread_rwlock_unlock(&st->mu);
    }

    /* war-mode guard: blocked users replying to the bot get purged */
    if (blocked && war) {
        /* resolve the replied-to message (inline or fetched) */
        uint64_t ref_author = 0;
        char *ref_content = NULL;
        if (m->has_ref_msg) {
            ref_author = m->ref_msg_author_id;
            ref_content = m->ref_msg_content ? xstrdup(m->ref_msg_content)
                                             : xstrdup("");
        } else if (m->has_reference) {
            disc_message_t orig;
            memset(&orig, 0, sizeof orig);
            if (discord_get_message(c, m->ref_channel_id, m->ref_message_id,
                                    &orig) == 0) {
                ref_author = orig.author_id;
                ref_content = orig.content ? xstrdup(orig.content)
                                           : xstrdup("");
                disc_message_free(&orig);
            }
        }
        if (ref_content) {
            if (ref_author == me) {
                discord_delete_message(c, m->channel_id, m->id);
                if (strcmp(ref_content, taunt) != 0)
                    discord_send_message(c, m->channel_id, taunt, NULL, 0,
                                         NULL);
            }
            free(ref_content);
        }
        return;
    }

    if (m->author_id == me)
        return;

    /* slash/prefix commands */
    {
        char name[64];
        const char *arg = NULL;
        if (cmd_parse_prefix(m->content ? m->content : "", name, sizeof name,
                             &arg) == 0) {
            handle_prefix(c, st, m);
            return;
        }
    }

    /* @artixy AI chat */
    if (ambient_mention(c, st, m))
        return;

    /* live-session typing: plain reply to a live message */
    if (!m->n_attachments && m->content && *m->content) {
        uint64_t target = 0;
        if (m->has_ref_msg && m->ref_msg_id)
            target = m->ref_msg_id;
        else if (m->has_reference)
            target = m->ref_message_id;
        if (target && st->live) {
            uint64_t owner_id = 0;
            char *fifo = NULL;
            if (live_session_for_msg(st->live, m->channel_id, target,
                                     &owner_id, &fifo)) {
                bool handled = true;
                if (owner_id != m->author_id) {
                    discord_delete_message(c, m->channel_id, m->id);
                    uint64_t dm = 0;
                    if (discord_create_dm(c, m->author_id, &dm) == 0 && dm)
                        discord_send_message(
                            c, dm,
                            "That live session belongs to someone else — "
                            "typing into it is blocked.",
                            NULL, 0, NULL);
                } else if (!fifo) {
                    discord_delete_message(c, m->channel_id, m->id);
                    uint64_t dm = 0;
                    if (discord_create_dm(c, m->author_id, &dm) == 0 && dm)
                        discord_send_message(
                            c, dm,
                            "That live session isn't interactive (fifo "
                            "missing).",
                            NULL, 0, NULL);
                } else if (!bot_is_authed(st, m->author_id)) {
                    handled = false;
                } else {
                    char trimmed[2048];
                    snprintf(trimmed, sizeof trimmed, "%s", m->content);
                    size_t tl = strlen(trimmed);
                    while (tl > 0 && (trimmed[tl - 1] == ' ' ||
                                      trimmed[tl - 1] == '\t' ||
                                      trimmed[tl - 1] == '\n'))
                        trimmed[--tl] = '\0';
                    if (!tl) {
                        free(fifo);
                        return;
                    }
                    char *payload = live_expand_typed_input(trimmed);
                    char vmname[256] = "";
                    char runas[64] = "";
                    {
                        pthread_rwlock_rdlock(&st->mu);
                        snprintf(vmname, sizeof vmname, "%s", st->vm);
                        pthread_rwlock_unlock(&st->mu);
                    }
                    const char *lx =
                        bot_linked_user(st, m->author_id);
                    if (bot_valid_runas(lx))
                        snprintf(runas, sizeof runas, "%s", lx);
                    bool ok = payload && live_forward_input(
                                           vmname, fifo,
                                           runas[0] ? runas : NULL, payload);
                    free(payload);
                    discord_delete_message(c, m->channel_id, m->id);
                    if (!ok) {
                        uint64_t dm = 0;
                        if (discord_create_dm(c, m->author_id, &dm) == 0 && dm)
                            discord_send_message(
                                c, dm,
                                "Couldn't type into that session (it just "
                                "ended).",
                                NULL, 0, NULL);
                    }
                }
                free(fifo);
                if (handled)
                    return;
            }
        }
    }

    if (!owner_admin) {
        if (war && events_is_boo(m->content ? m->content : "")) {
            discord_send_reply(c, m->channel_id, m->id, "boo on you! :3", NULL,
                               0, NULL);
        }
        return;
    }

    /* sayas-auto: plain owner/admin messages are reposted as artix */
    if (sayas_auto) {
        const char *content = m->content ? m->content : "";
        bool has_files = m->n_attachments > 0;
        /* trim leading check */
        while (*content == ' ' || *content == '\t')
            content++;
        char *ar_probe = events_artixy_text(m->content);
        bool is_ar = ar_probe != NULL;
        free(ar_probe);
        bool plain_text = *content && content[0] != '/' && content[0] != ';' &&
                          !is_ar;
        if (plain_text || (has_files && !*content)) {
            disc_file_t *files = NULL;
            size_t n_files = 0;
            char **bufs = NULL;
            if (has_files &&
                bot_download_atts(m->attachments, m->n_attachments, &files,
                                  &n_files, &bufs) != 0) {
                files = NULL;
                n_files = 0;
                bufs = NULL;
            }
            char body[2100];
            snprintf(body, sizeof body, "%s", content);
            if (strlen(body) > 2000) {
                /* overflow -> file (same shape as sayas) */
                char *clean = bot_strip_sgr(body);
                size_t need = (clean ? strlen(clean) : 0) + 1;
                char *fdata = xmalloc(need);
                disc_file_t *nf = NULL;
                char **nb = NULL;
                if (fdata) {
                    snprintf(fdata, need, "%s", clean ? clean : "");
                    nf = xrealloc(files, (n_files + 1) * sizeof *nf);
                    nb = xrealloc(bufs, (n_files + 1) * sizeof *nb);
                }
                free(clean);
                if (fdata && nf && nb) {
                    files = nf;
                    bufs = nb;
                    files[n_files].name = xstrdup("output.txt");
                    files[n_files].data = fdata;
                    files[n_files].len = strlen(fdata);
                    bufs[n_files] = fdata;
                    if (files[n_files].name)
                        n_files++;
                    else
                        free(fdata);
                } else {
                    free(fdata);
                }
                body[0] = '\0';
            }
            if (!body[0] && !n_files) {
                bot_free_dl_files(files, n_files, bufs);
                return;
            }
            discord_delete_message(c, m->channel_id, m->id);
            if (body[0])
                ai_record_artixy(m->channel_id, body);
            uint64_t reply_to = 0;
            if (m->has_ref_msg && m->ref_msg_id)
                reply_to = m->ref_msg_id;
            else if (m->has_reference)
                reply_to = m->ref_message_id;
            post_as_artix(c, m->channel_id, reply_to, body, files, n_files);
            bot_free_dl_files(files, n_files, bufs);
            return;
        }
    }

    /* trailing ".ar" posts the line as artix */
    {
        char *text = events_artixy_text(m->content ? m->content : "");
        if (!text)
            return;
        for (int i = 0; i < 3; i++) {
            if (discord_delete_message(c, m->channel_id, m->id) == 0)
                break;
            fprintf(stderr, "artixy-say: delete attempt failed for %llu\n",
                    (unsigned long long)m->id);
            struct timespec ts = { 1, 0 };
            nanosleep(&ts, NULL);
        }
        disc_file_t *files = NULL;
        size_t n_files = 0;
        char **bufs = NULL;
        if (m->n_attachments)
            bot_download_atts(m->attachments, m->n_attachments, &files,
                              &n_files, &bufs);
        char body[2100];
        snprintf(body, sizeof body, "%s", text);
        free(text);
        if (strlen(body) > 2000) {
            char *clean = bot_strip_sgr(body);
            size_t need = (clean ? strlen(clean) : 0) + 1;
            char *fdata = xmalloc(need);
            if (fdata) {
                snprintf(fdata, need, "%s", clean ? clean : "");
                disc_file_t *nf =
                    xrealloc(files, (n_files + 1) * sizeof *nf);
                char **nb = xrealloc(bufs, (n_files + 1) * sizeof *nb);
                if (nf && nb) {
                    files = nf;
                    bufs = nb;
                    files[n_files].name = xstrdup("output.txt");
                    files[n_files].data = fdata;
                    files[n_files].len = strlen(fdata);
                    bufs[n_files] = fdata;
                    if (files[n_files].name)
                        n_files++;
                    else
                        free(fdata);
                } else {
                    free(fdata);
                }
            }
            free(clean);
            body[0] = '\0';
        }
        if (body[0])
            ai_record_artixy(m->channel_id, body);
        uint64_t reply_to = 0;
        if (m->has_ref_msg && m->ref_msg_id)
            reply_to = m->ref_msg_id;
        else if (m->has_reference)
            reply_to = m->ref_message_id;
        if (body[0] || n_files)
            post_as_artix(c, m->channel_id, reply_to, body, files, n_files);
        bot_free_dl_files(files, n_files, bufs);
    }
}
